src/game/BattleGround/BattleGroundWS.cpp

```diff
@@ -217,218 +293,143 @@ void BattleGroundWS::EventPlayerCapturedFlag(Player* source)
-            }
-        }
-        else
-        {
-            if (!IsAllianceFlagPickedUp())
-                return;
-            if (GetAllianceFlagCarrierGuid() == source->GetObjectGuid())
-            {
-                ClearAllianceFlagCarrier();
-                source->RemoveAurasDueToSpell(BG_WS_SPELL_SILVERWING_FLAG);
-            }
-        }
-        return;
-    }
-
-    bool set = false;
+    PvpTeamIndex teamIdx = GetTeamIndexByTeamId(player->GetTeam());
+    PvpTeamIndex otherTeamIdx = GetOtherTeamIndex(teamIdx);
 
-    if (source->GetTeam() == ALLIANCE)
+    if (GetStatus() != STATUS_IN_PROGRESS)
     {
-        if (!IsHordeFlagPickedUp())
-            return;
-        if (GetHordeFlagCarrierGuid() == source->GetObjectGuid())
+        if (IsFlagPickedUp(teamIdx) && GetFlagCarrierGuid(teamIdx) == player->GetObjectGuid())
         {
-            ClearHordeFlagCarrier();
-            source->RemoveAurasDueToSpell(BG_WS_SPELL_WARSONG_FLAG);
-            m_FlagState[TEAM_INDEX_HORDE] = BG_WS_FLAG_STATE_ON_GROUND;
-            source->CastSpell(source, BG_WS_SPELL_WARSONG_FLAG_DROPPED, TRIGGERED_OLD_TRIGGERED);
-            set = true;
+            ClearFlagCarrier(teamIdx);
+            player->RemoveAurasDueToSpell(wsSpellTypes[teamIdx][BG_WS_FLAG_ACTION_PICKEDUP]);
         }
     }
-    else
+    else if (IsFlagPickedUp(otherTeamIdx))
     {
-        if (!IsAllianceFlagPickedUp())
-            return;
-        if (GetAllianceFlagCarrierGuid() == source->GetObjectGuid())
+        if (GetFlagCarrierGuid(otherTeamIdx) == player->GetObjectGuid())
         {
-            ClearAllianceFlagCarrier();
-            source->RemoveAurasDueToSpell(BG_WS_SPELL_SILVERWING_FLAG);
-            m_FlagState[TEAM_INDEX_ALLIANCE] = BG_WS_FLAG_STATE_ON_GROUND;
-            source->CastSpell(source, BG_WS_SPELL_SILVERWING_FLAG_DROPPED, TRIGGERED_OLD_TRIGGERED);
-            set = true;
+            ClearFlagCarrier(otherTeamIdx);
+            player->RemoveAurasDueToSpell(wsSpellTypes[otherTeamIdx][BG_WS_FLAG_ACTION_PICKEDUP]);
+            m_FlagState[otherTeamIdx] = BG_WS_FLAG_STATE_ON_GROUND;
+            player->CastSpell(player, wsSpellTypes[otherTeamIdx][BG_WS_FLAG_ACTION_DROPPED], TRIGGERED_OLD_TRIGGERED);
         }
-    }
 
-    if (set)
-    {
-        source->CastSpell(source, SPELL_RECENTLY_DROPPED_FLAG, TRIGGERED_OLD_TRIGGERED);
-        UpdateFlagState(source->GetTeam(), 1);
-
-        if (source->GetTeam() == ALLIANCE)
-        {
-            SendMessageToAll(LANG_BG_WS_DROPPED_HF, CHAT_MSG_BG_SYSTEM_HORDE, source);
-            UpdateWorldState(BG_WS_FLAG_UNK_HORDE, uint32(-1));
-        }
-        else
-        {
-            SendMessageToAll(LANG_BG_WS_DROPPED_AF, CHAT_MSG_BG_SYSTEM_ALLIANCE, source);
-            UpdateWorldState(BG_WS_FLAG_UNK_ALLIANCE, uint32(-1));
-        }
+        player->CastSpell(player, SPELL_RECENTLY_DROPPED_FLAG, TRIGGERED_OLD_TRIGGERED);
+        UpdateFlagState(player->GetTeam(), 1);
```

Comment:
GetTeam() should be cached at start of function and used in all cases where GetTeam is called